Tests/MongoSwiftTests/RetryableWritesTests.swift

```diff
@@ -0,0 +1,167 @@
+/// Struct representing a single test within a spec test JSON file.
+private struct RetryableWritesTest: Decodable, SpecTest {
+    let description: String
+    let outcome: TestOutcome
+    let operation: AnyTestOperation
+
+    /// Options used to configure the `MongoClient` used for this test.
+    let clientOptions: ClientOptions?
+
+    /// If true, the `MongoClient` for this test should be initialized with multiple mongos seed addresses.
+    /// If false or omitted, only a single mongos address should be specified.
+    /// This field has no effect for non-sharded topologies.
+    let useMultipleMongoses: Bool?
+
+    /// The optional configureFailPoint command document to run to configure a fail point on the primary server.
+    /// This option and useMultipleMongoses: true are mutually exclusive.
+    let failPoint: Document?
+}
+
+/// Struct representing a single retryable-writes spec test JSON file.
+private struct RetryableWritesTestFile: Decodable {
+    private enum CodingKeys: CodingKey {
+        case runOn, data, tests
+    }
+
+    /// Name of this test case
+    var name: String = ""
+
+    /// Server version and topology requirements in order for tests from this file to be run.
+    let runOn: [TestRequirement]?
+
+    /// Data that should exist in the collection before running any of the tests.
+    let data: [Document]
+
+    /// List of tests to run in this file.
+    let tests: [RetryableWritesTest]
+}
+
+final class RetryableWritesTests: MongoSwiftTestCase {
+    /// If a failpoint was set on the current test being run, the command in this document will disable it.
+    var disableFailPointCommand: Document?
+
+    override func tearDown() {
+        if let cmd = self.disableFailPointCommand {
+            do {
+                let client = try MongoClient(MongoSwiftTestCase.connStr)
+                try client.db("admin").runCommand(cmd)
+            } catch {
+                print("couldn't disable failpoints after failure: \(error)")
+            }
+        }
+    }
+
+    override func setUp() {
+        self.continueAfterFailure = false
+    }
+
+    // Teardown at the very end of the suite by dropping the db we tested on.
+    override class func tearDown() {
+        super.tearDown()
+        do {
+            try MongoClient().db(self.testDatabase).drop()
+        } catch {
+            print("Dropping test db \(self.testDatabase) failed: \(error)")
+        }
+    }
+
+    func testRetryableWrites() throws {
+        let testFilesPath = MongoSwiftTestCase.specsPath + "/retryable-writes/tests"
+        let testFiles: [String] = try FileManager.default.contentsOfDirectory(atPath: testFilesPath)
+
+        let tests: [RetryableWritesTestFile] = try testFiles.map { fileName in
+            let url = URL(fileURLWithPath: "\(testFilesPath)/\(fileName)")
+            let data = try String(contentsOf: url).data(using: .utf8)!
+            var testFile = try JSONDecoder().decode(RetryableWritesTestFile.self, from: data)
```

Comment:
since we are decoding various documents, it might be more performant to use the `BSONDecoder` here rather than `JSONDecoder`, since using `JSONDecoder` will necessitate decoding each value in a doc individually via `AnyBSONValue` which uses that long switch statement to test every type.
OTOH we have minimal tests for how `JSONDecoder` works with `Document`, so maybe it's good to do that for once 🤷‍♀